### PR TITLE
Optimise new submission creation

### PR DESCRIPTION
**Pull request by ekowidianto**
When a user attempts a submission for the first time, submission_questions and answers are created and inserted 1 by 1 to the DB resulting in N+1 issue. To solve this, we now use bulk insert instead.

In addition, a new submission of a randomized assessment would create answers and submission_questions for all questions from the assessment. Now, only assigned questions based on the question bundle assignmenets are used.

Improvement comparisons (create new attempt on courses/1745/assessments/30193):
- Local (before: ~2600ms, after: ~1300ms)
- Staging (before: ~700ms, after: ~ms)
- Production (before: ~1000ms, after: ~ms)

In detail (local):
Before
image

After
image

### CODE REVIEW

**Comment by ekowidianto** [comment]
@zhuhanming Asking your opinion for this issue if we could do it this way or any other alternatives. With the bulk insert for the submission_question, we could cut another ~20% of the time to execute the whole thing. Minor downside is that since submission_question is an actable of discussion post, we need to explicitly do another bulk insert for the post. CMIIW but currently there is no way to do bulk insert with association.

It is also possible to do the same thing for the answers creation ([def load_or_create_answers](https://github.com/Coursemology/coursemology2/blob/3c84ce4f9250511835fbe1f7a74909e43b1896d4/app/services/course/assessment/submission/update_service.rb#L34)). I think it may be similar to the submission_question above, where we need to bulk insert the answer actables > then bulk insert answers, but it gets a little bit trickier here since some answer actables have another association to save (eg for programming answer actables > programming files). So it gets a little bit more manual here to figure out all the associations that are saved and we need to write the bulk insert separately for each table. Also, since we have 6 different types of the answer actables, there will be different types of question in an assessment and the bulk insert will need to be done separately (best case all same then 1 bulk insert or worst case all different then nothing happens). What do you think about this? Or we could just stick to the current way of saving (i.e. individual save)

TLDR; working with bulk insert for MTI is tricky.

File: app/models/concerns/course/assessment/submission/answers_concern.rb [hunk 1]
```diff
@@ -17,15 +17,37 @@ def from_question(question_id)
...
+    new_answers_group_by_actables = new_answers_actables.group_by { |actable| actable.class.to_s }
 
-    new_answers.map do |answer|
-      # When there are no existing answers, the first one will be the current_answer.
-      if answer.new_record?
-        answer.current_answer = true
-        answer.save!
+    ActiveRecord::Base.transaction do
+      new_answers_group_by_actables.each_key do |key|
+        key.constantize.import! new_answers_group_by_actables[key], recursive: true
```
Comment (ekowidianto) [inline app/models/concerns/course/assessment/submission/answers_concern.rb:48]: The gem activerecord-import (https://github.com/zdennis/activerecord-import) enables bulk insert without skipping validation and allowing associated new records of different tables to be inserted as well.

In Rails 6, insert_all and upsert_all are added but the functionalities above are not available yet.

File: app/services/course/assessment/submission/update_service.rb [hunk 2]
```diff
@@ -138,22 +130,25 @@ def create_missing_submission_questions
     new_submission_questions = []
     questions_without_submission_questions.each do |question|
       new_submission_questions <<
-        @submission.submission_questions.build(question: question)
+        Course::Assessment::SubmissionQuestion.new(submission: @submission, question: question)
```
Comment (ekowidianto) [inline app/services/course/assessment/submission/update_service.rb:133]: Instead of using build, we use new to create new records here to allow the submission_question acting_as (discussion post) to be bulk inserted using activerecord-import

**Comment by ekowidianto** [comment]
Comparisons for randomized assessment (2040S Final)

**Before**
image

**After**
image

File: app/models/concerns/course/assessment/submission/answers_concern.rb [hunk 3]
```diff
@@ -17,15 +17,38 @@ def from_question(question_id)
   end
 
   def create_new_answers
-    questions_to_attempt ||= assessment.questions
+    # Load questions from submission instead of assessment in case of randomized assessment
+    questions_to_attempt ||= questions.includes(:actable)
```
Comment (zhuhanming) [inline app/models/concerns/course/assessment/submission/answers_concern.rb:21]: Can I check where this `questions` is coming from?

**Review by zhuhanming — APPROVED** [review]
Nice, looks good to me! Will test it out locally.

File: app/models/concerns/course/assessment/submission/answers_concern.rb (hunk 3 above)
Comment (ekowidianto) [reply app/models/concerns/course/assessment/submission/answers_concern.rb:21]: Its defined here:

https://github.com/Coursemology/coursemology2/blob/79f814349ad54ba571533921eff331782eba9fbd/app/models/course/assessment/submission.rb#L185-L187
